storage/store_mgr.go

```diff
@@ -15,79 +15,246 @@
+
+// Check if all stores types are provided by stores config
+// If no config is provided, global default config store is added
+// If any store type is missing, a related default config store is added
+func checkAndBuildConfigStores(cfg config.Configuration) []config.Store {
+	if cfg.Storage.Stores == nil {
+		return buildDefaultConfigStores()
+	}
+
+	cfgStores := cfg.Storage.Stores
+	checkStores := make([]string, 0)
+	for _, configStore := range cfg.Storage.Stores {
+		for _, storeTypeName := range configStore.Types {
+			// let's do this case insensitive
+			name := strings.ToLower(storeTypeName)
+			if !collections.ContainsString(checkStores, name) {
+				checkStores = append(checkStores, name)
 			}
-			stores[typ] = store
 		}
 	}
 
-	for _, typ := range types.StoreTypeNames() {
-		st, _ := types.ParseStoreType(typ)
-		if _, ok := stores[st]; !ok {
-			log.Printf("Using default store for type: %q.", typ)
-			stores[st] = defaultStore
+	for _, storeTypeName := range types.StoreTypeNames() {
+		name := strings.ToLower(storeTypeName)
+		if !collections.ContainsString(checkStores, name) {
+			log.Printf("Default config store will be used for store type:%q.", storeTypeName)
+			var defaultStore config.Store
+			switch storeTypeName {
+			case types.StoreTypeDeployment.String():
+				defaultStore = config.Store{
+					Name:           "defaultFileStoreWithCache",
+					Implementation: fileStoreWithCacheImpl,
+					Types:          []string{types.StoreTypeDeployment.String()},
+				}
+			case types.StoreTypeEvent.String():
+				defaultStore = config.Store{
+					Name:           "defaultConsulStore" + types.StoreTypeEvent.String(),
+					Implementation: consulStoreImpl,
+					Types:          []string{types.StoreTypeEvent.String()},
+				}
+			case types.StoreTypeLog.String():
+				defaultStore = config.Store{
+					Name:           "defaultConsulStore" + types.StoreTypeLog.String(),
+					Implementation: consulStoreImpl,
+					Types:          []string{types.StoreTypeLog.String()},
+				}
+			}
+
+			cfgStores = append(cfgStores, defaultStore)
 		}
 	}
+	return cfgStores
+}
 
+// Create store implementations
+func createStoreImpl(cfg config.Configuration, configStore config.Store, storeType types.StoreType) error {
+	var err error
+	impl := strings.ToLower(configStore.Implementation)
+	switch impl {
+	case strings.ToLower(fileStoreWithCacheImpl), strings.ToLower(fileStoreWithCacheAndEncryptionImpl):
+		stores[storeType], err = file.NewStore(cfg, configStore.Name, configStore.Properties, true, impl == strings.ToLower(fileStoreWithCacheAndEncryptionImpl))
+		if err != nil {
+			return err
+		}
+	case strings.ToLower(consulStoreImpl):
+		stores[storeType] = consul.NewStore()
+	default:
+		log.Printf("[WARNING] unknown store implementation:%q. This will be ignored.", storeType)
+	}
 	return nil
 }
 
 // GetStore returns the store related to a defined store type
-func GetStore(typ types.StoreType) Store {
+func GetStore(typ types.StoreType) store.Store {
```

Comment:
Maybe use check existence form (the one that also returns a bool) of the get key in map to validate WDYT?

Reply:
Makes sense

Reply:
But at this point, as each store is mandatory, we should panic if the store is missing

Reply:
Or just return the bool if later we handle optional stores ?

Reply:
But it makes the code more complex for the caller:
``

Reply:
https://github.com/golang/go/wiki/CodeReviewComments#dont-panic

Reply:
This is not a "normal" error as the stores have to exist. Panic is relevant in this case.

In the future, if we handle optional store, we may have to change this but not sure to need it ?

Reply:
:+1: